doc/design/context.md

```diff
@@ -0,0 +1,166 @@
+## Context Design
+
+
+A Net is executed by single or several threads. A Context is related to a thread and records necessary runtime resources.
+
+Context is defined as follows:
+
+```
+class Context {};
```

Comment:
no common methods shared between GpuContext and CpuContext? 

```c++
struct Context {
    int dev_id{0}; // CPU = 0
    enum DevType {
        kCPU;
        kGPU;
    };
   DevType dev_type;
   
   enum StreamType {
       kCUDNN;
       kBLAS;
       kCUDA;
   };
  // all the streams are created globally, so a void* is enough
  // idx is the index of stream of `type`, each thread in a device can have one idx.
  void* GetStream(StreamType type, int stream_idx);
  // allocate one stream for each StreamType and thread, in all the devices.
  static void** streams;
};
```

Both `CpuContext` and `GpuContext` are `Context`, so it's wired that they have no similarity.

And we can merge  them into one `Context` , really need two? 
If the `CpuContext` and `GpuContext` will be used as a template parameter, can `Place` repalce their role ?